### PR TITLE
[4.0] Remove Meta Keywords (part 1)

### PR DESCRIPTION
**Pull request by brianteeman**
Meta Keywords as a meta tag has pretty much been irrelevant for **10 years** (see https://yoast.com/meta-keywords/  for example)

This PR does the following

1. Removes Site Meta Keywords from global config
2. Removes Menu item Meta Keywords
3. Removes per language Meta Keywords
4. Renames the item field Meta Keywords to Keywords
 - We use this as a field with mod_related_items and com_banners

Part 2 of removing meta keywords will
1. Move the Keyword field to a more appropriate place
2. Check that all the keyword fields are used with mod_related_items and com_banners

### Testing
Check that Meta Keyword field is removed from global config and menu items
Check that renamed Keyword field does not produce a meta key in the frontend

### CODE REVIEW

File: libraries/src/Table/CoreContent.php [hunk 1]
```diff
@@ -145,36 +145,6 @@ public function check()
...
-		// and cr (\r) and lf (\n) characters from string
-		if (!empty($this->core_metakey))
-		{
-			// Only process if not empty
-
-			// Array of characters to remove
-			$bad_characters = array("\n", "\r", "\"", '<', '>');
-
-			// Remove bad characters
-			$after_clean = StringHelper::str_ireplace($bad_characters, '', $this->core_metakey);
```
Comment (Quy) [inline libraries/src/Table/CoreContent.php]: Remove `use Joomla\String\StringHelper;`

File: libraries/src/Table/Content.php [hunk 2]
```diff
@@ -272,36 +272,6 @@ public function check()
...
-		// and cr (\r) and lf (\n) characters from string
-		if (!empty($this->metakey))
-		{
-			// Only process if not empty
-
-			// Array of characters to remove
-			$bad_characters = array("\n", "\r", "\"", '<', '>');
-
-			// Remove bad characters
-			$after_clean = StringHelper::str_ireplace($bad_characters, '', $this->metakey);
```
Comment (Quy) [inline libraries/src/Table/Content.php]: Remove `use Joomla\String\StringHelper;`

**Comment by HLeithner** [comment]
Why do you think we should remove it? Who says it's irrelevant in all cases? Is this true for all search engines or only for google?

Our own search engine (smart search) uses the metakey for the ranking...

**Comment by brianteeman** [comment]
@HLeithner 
>Why do you think we should remove it? Who says it's irrelevant in all cases? Is this true for all search engines or only for google?

https://yoast.com/meta-keywords/ neither yahoo, google or bing use it

> Our own search engine (smart search) uses the metakey for the ranking...

As stated in the PR
 > Renames the item field Meta Keywords to Keywords

**Comment by brianteeman** [comment]
@HLeithner This was also part of the specification for the GSOC 19 SEO project https://docs.google.com/document/d/1MaOtEpp0Ma-e_d7xuXVAnGXP72ec9Ar1XrKtPQvnw4I/edit?usp=sharing

**Comment by HLeithner** [comment]
Adding the rest would be much nicer then removing something thats maybe not used ;-) but ok if george or the other maintainers are happy with it i'm too.

**Comment by brianteeman** [comment]
> Adding the rest would be much nicer

I am working on some of those too

yourfreedomcoach-Mervyn-Barrett-constant-improvement

**Comment by simbus82** [comment]
Yoast is the worst source for "SEO best practice". Rubbish for wordpress newbies.

https://yandex.com/support/webmaster/controlling-robot/html.html
https://www.searchenginejournal.com/yandex-seo-guide/252885/

Meta Tag Keywords are important for SEO in Russia.
image

It is not a problem to use some plugin or some custom fields to obtain this for russian websites.
And I am the first to tell my clients not to waste time filling in that data. But only to Italian, English, French, etc. customers ... Not to Russian customers!
But Joomla needs a SEO team leader who should have a little more experience in international projects. By the way ... is there a SEO team leader in the organization?

**Comment by alikon** [comment]
> But Joomla needs a SEO team leader who should have a little more experience in international projects. 

@simbus82  i understand this like a candidature to do this work, 
and imho we really need people that want to do work .....   

so please submit a PR or a RFC ....whatever

p.s.
italians do it better

**Comment by brianteeman** [comment]
The reason for posting the toast link was to save time posting one link and not three. I am just submitting a pr to match the reques I have no clue on SEO.

File: administrator/language/en-GB/en-GB.com_banners.ini [hunk 3]
```diff
@@ -74,7 +74,7 @@ COM_BANNERS_FIELD_BASENAME_LABEL="File Name"
 COM_BANNERS_FIELD_CLICKS_LABEL="Total Clicks"
 COM_BANNERS_FIELD_CLICKURL_LABEL="Click URL"
 COM_BANNERS_FIELD_CLIENT_LABEL="Client"
-COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_DESC="When matching Meta keywords, only search for Meta keywords with this prefix (improves performance)."
+COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_DESC="When matching Keywords, only search for Keywords with this prefix (improves performance)."
 COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_LABEL="Meta Keyword Prefix"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_banners.ini]:
```suggestion
COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_LABEL="Keyword Prefix"
```

File: administrator/language/en-GB/en-GB.com_banners.ini [hunk 4]
```diff
@@ -89,7 +89,7 @@ COM_BANNERS_FIELD_HEIGHT_LABEL="Height"
 COM_BANNERS_FIELD_IMAGE_LABEL="Image"
 COM_BANNERS_FIELD_IMPMADE_LABEL="Total Impressions"
 COM_BANNERS_FIELD_IMPTOTAL_LABEL="Max. Impressions"
-COM_BANNERS_FIELD_METAKEYWORDPREFIX_DESC="When matching Meta keywords, only search for Meta keywords with this prefix (improves performance)."
+COM_BANNERS_FIELD_METAKEYWORDPREFIX_DESC="When matching Keywords, only search for Keywords with this prefix (improves performance)."
 COM_BANNERS_FIELD_METAKEYWORDPREFIX_LABEL="Meta Keyword Prefix"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_banners.ini]:
```suggestion
COM_BANNERS_FIELD_METAKEYWORDPREFIX_LABEL="Keyword Prefix"
```

File: language/en-GB/en-GB.mod_banners.ini [hunk 5]
```diff
@@ -13,7 +13,7 @@ MOD_BANNERS_FIELD_COUNT_LABEL="Count"
 MOD_BANNERS_FIELD_FOOTER_LABEL="Footer Text"
 MOD_BANNERS_FIELD_HEADER_LABEL="Header Text"
 MOD_BANNERS_FIELD_RANDOMISE_LABEL="Randomise"
-MOD_BANNERS_FIELD_TAG_DESC="Banner is selected by matching the banner tags to the current document meta keywords."
+MOD_BANNERS_FIELD_TAG_DESC="Banner is selected by matching the banner tags to the current document Keywords."
 MOD_BANNERS_FIELD_TAG_LABEL="Select by Meta Keyword"
```
Comment (Quy) [inline language/en-GB/en-GB.mod_banners.ini]:
```suggestion
MOD_BANNERS_FIELD_TAG_LABEL="Select by Keyword"
```

File: libraries/src/Table/CoreContent.php [hunk 6]
```diff
@@ -145,36 +144,6 @@ public function check()
 			$this->core_publish_down = $temp;
 		}
 
-		// Clean up keywords -- eliminate extra spaces between phrases
```
Comment (wilsonge) [inline libraries/src/Table/CoreContent.php]: This one we probably still need - anything 3rd party using UCM might be using this (at least JED) and whilst we can make the b/c break I'm not sure it's *that* important

File: libraries/src/Table/Content.php [hunk 7]
```diff
@@ -272,36 +271,6 @@ public function check()
 			$this->modified = $this->_db->getNullDate();
 		}
 
-		// Clean up keywords -- eliminate extra spaces between phrases
```
Comment (wilsonge) [inline libraries/src/Table/Content.php]: This is the Table instance for com_content - we're keeping the key's here for the keywords field so this is required for sure

**Comment by wilsonge** [comment]
I've edited the yoast url into the PR description so anyone coming to this PR sees it first :)

File: libraries/src/Table/Content.php (hunk 7 above)
Comment (brianteeman) [reply libraries/src/Table/Content.php]: Do we really need the cleanup? Happy to put it back if we do

**Comment by wilsonge** [comment]
@simbus82 you don't need meta keywords for yandex if you're using json ld which we will be once I finish up https://github.com/joomla/joomla-cms/pull/25117 (yandex has been supporting that program since 2011)

File: libraries/src/Table/Content.php (hunk 7 above)
Comment (wilsonge) [reply libraries/src/Table/Content.php]: I think it's safer. Because otherwise we're going to end up having issues with search which expects the data in a certain format

File: libraries/src/Table/CoreContent.php (hunk 6 above)
Comment (brianteeman) [reply libraries/src/Table/CoreContent.php]: done

File: libraries/src/Table/Content.php (hunk 7 above)
Comment (brianteeman) [reply libraries/src/Table/Content.php]: done

**Comment by wilsonge** [comment]
We're going to need to document a b/c break to banners if/when we go with this. It seems to filter by the documents keywords. So things like the menu and global config metadata as fallbacks will be relevant there. https://github.com/joomla/joomla-cms/blob/staging/modules/mod_banners/helper.php#L35

In the mean time I'm going to try and ping some people in the Russian community to gain some extra clarification over Yandex

**Comment by brianteeman** [comment]
Re Yandex it does use keywords if this is up to date https://yandex.com/support/webmaster/controlling-robot/html.html

**Comment by atpatil11** [comment]
Getting the following error while applying patch.

The file marked for modification does not exist: components/com_search/View/Search/HtmlView.phpscreen shot 2019-11-02 at 06 22 09<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

**Comment by Quiviro** [comment]
I have tested this item :red_circle: unsuccessfully on a6409db7d40cef895d8b5e1013c5036897ee0257<br /><br />Same error message than [alpatil11](https://issues.joomla.org/tracker/joomla-cms/25258#event-485633 "").
Meta Keywords field still in global config and menu item<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

**Comment by b2z** [comment]
Correct me if I am wrong. Updating from J3 to J4 will affect only this - it won't produce a meta key in the frontend?

A little poll by me in Russian Telegram chat about keywords states that keywords are 99.9% useless for Yandex.

**Comment by wilsonge** [comment]
Correct

**Comment by brianteeman** [comment]
So do I _spend my time_ updating this pr to resolve the conflicts or not?

**Comment by b2z** [comment]
Here is one experience with keywords and Yandex. Quoted and translated from Russian:

For a long time I was convinced that keyword is a rudiment, about a year ago I read an article about an experiment where a person added phrases to keywords that were not on the page, and after that he received visibility through them after the  time. Believe it or not - this is, as usual in our topic, everyone’s business, but I personally leave it to myself.

**Comment by Quy** [comment]
Unable to install your branch.

**Comment by brianteeman** [comment]
I think I messed up fixing the conflicts

**Comment by b2z** [comment]
Just leaving a note - people keep saying that keywords for Yandex is almost useless and for those who want to keep them, can create a plugin :)

**Comment by 4digitalworld** [comment]
@b2z Maybe it will be a **official plugin**, from the main Joomla directory, as it was done for the [weblinks](https://extensions.joomla.org/extension/official-extensions/weblinks/) component?

@brianteeman I mean, it can be moved as a separate plugin here: [https://extensions.joomla.org/category/official-extensions/](https://extensions.joomla.org/category/official-extensions/)

**Comment by b2z** [comment]
Well, IMO keywords does not deserve such attention in 2020.

**Comment by brianteeman** [comment]
@sanek4life no - either keep it or kill it

File: installation/sql/postgresql/joomla.sql [hunk 8]
```diff
@@ -1368,7 +1368,7 @@ SELECT 19, 'main', 'mod_menu_fields', 'Contact Custom Fields', '', 'contact/Cust
...
-SELECT 101, 'mainmenu', 'Home', 'home', '', 'home', 'index.php?option=com_content&view=featured', 'component', 1, 1, 1, "extension_id", 0, NULL, 0, 1, '', 0, '{"featured_categories":[""],"layout_type":"blog","num_leading_articles":"1","num_intro_articles":"3","num_columns":"3","num_links":"0","multi_column_order":"1","orderby_pri":"","orderby_sec":"front","order_date":"","show_pagination":"2","show_pagination_results":"1","show_title":"","link_titles":"","show_intro":"","info_block_position":"","show_category":"","link_category":"","show_parent_category":"","link_parent_category":"","show_author":"","link_author":"","show_create_date":"","show_modify_date":"","show_publish_date":"","show_item_navigation":"","show_vote":"","show_readmore":"","show_readmore_title":"","show_hits":"","show_noauth":"","show_feed_link":"1","feed_summary":"","menu-anchor_title":"","menu-anchor_css":"","menu_image":"","menu_text":1,"page_title":"","show_page_heading":1,"page_heading":"","pageclass_sfx":"","menu-meta_description":"","menu-meta_keywords":"","robots":"","secure":0}', 39, 40, 1, '*', 0, NULL, NULL FROM "#__extensions" WHERE "name" = 'com_content';
+
```
Comment (wilsonge) [inline installation/sql/postgresql/joomla.sql]: assuming this wasn't supposed to be fully removed?

File: administrator/components/com_languages/forms/language.xml [hunk 9]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <form>
 	<fieldset>
-		<field
-			name="lang_id"
+		<field
```
Comment (wilsonge) [inline administrator/components/com_languages/forms/language.xml:4]: Revert this change please

File: libraries/src/Table/Content.php [hunk 10]
```diff
@@ -274,6 +274,12 @@ public function check()
 			$this->publish_down = $temp;
 		}
 
+		// Set modified to null date if not set
```
Comment (wilsonge) [inline libraries/src/Table/Content.php]: Not quite sure what went wrong here. But assume this isn't intentional?

**Comment by wilsonge** [comment]
Clearly some conflict problems but I'm happy we're not totally impeding the russian community here so let's proceed with this PR

**Comment by brianteeman** [comment]
might just make a new pr - easier than fixing the messed up merge and conflict

**Comment by brianteeman** [comment]
conflicts all resolved - now I understand cherry pick and force push ;)

File: libraries/src/Table/Content.php [hunk 11]
```diff
@@ -303,15 +309,6 @@ public function check()
 			// Put array back together delimited by ", "
 			$this->metakey = implode(', ', $clean_keys);
 		}
-		else
-		{
-			$this->metakey = '';
-		}
-
-		if ($this->metadesc === null)
```
Comment (Quy) [inline libraries/src/Table/Content.php]: Revert?

File: language/en-GB/mod_related_items.sys.ini [hunk 12]
```diff
@@ -5,4 +5,5 @@
...
+MOD_RELATED_XML_DESCRIPTION="This module displays other Articles that are related to the one being viewed. These relations are established by the Keywords. <br>All the keywords of the current Article are searched against all the keywords of all other published Articles. For example, you may have an Article on &quot;Breeding Parrots&quot; and another on &quot;Hand Raising Black Cockatoos&quot;. If you include the keyword &quot;parrot&quot; in both Articles, then the Related Items Module will list the &quot;Breeding Parrots&quot; Article when viewing &quot;Hand Raising Black Cockatoos&quot; and vice-versa."
+
```
Comment (Quy) [inline language/en-GB/mod_related_items.sys.ini]:
```suggestion
```

Comment (brianteeman) [reply language/en-GB/mod_related_items.sys.ini]: I am doing this to make you happy but almost all language files have an extra line

File: administrator/language/en-GB/com_banners.ini [hunk 13]
```diff
@@ -135,7 +135,7 @@ COM_BANNERS_HEADING_COUNT_DESC="Count descending"
 COM_BANNERS_HEADING_IMPRESSIONS="Impressions"
 COM_BANNERS_HEADING_IMPRESSIONS_ASC="Impressions ascending"
 COM_BANNERS_HEADING_IMPRESSIONS_DESC="Impressions descending"
-COM_BANNERS_HEADING_METAKEYWORDS="Meta Keywords"
+COM_BANNERS_HEADING_METAKEYWORDS="Keywords"
```
Comment (Quy) [inline administrator/language/en-GB/com_banners.ini]: Remove as it is not being used.

Comment (brianteeman) [reply administrator/language/en-GB/com_banners.ini]: was it being used before this PR?

Comment (Quy) [reply administrator/language/en-GB/com_banners.ini]: It doesn't appear to be. I did a search on the constant and found nothing.

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 2e575aa377a30577ada00e8547fd6548d6e49c22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2e575aa377a30577ada00e8547fd6548d6e49c22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

**Comment by brianteeman** [comment]
yay!!!!

**Comment by richard67** [comment]
@brianteeman Today at 12 UTC:
j4-install-persian-language-notice

Content language isn't created with the language installation.

**Comment by richard67** [comment]
New issue or PR?

**Comment by richard67** [comment]
Will work on it.

**Comment by richard67** [comment]
Is SQL problem ... will fix it, be patient for new PR.

**Comment by richard67** [comment]
@brianteeman Shall the column `keywords` be removed from database table `#__languages`, or shall we keep it and use empty string default value? And how about other places where kewords were removed? I.e. what is the plan?

**Comment by brianteeman** [comment]
the intention was not to remove the fields in this pr as you will see that the field is still used for mod_related and com_banners

**Comment by richard67** [comment]
@brianteeman Thanks for feedback. Yes, noticed that, too, and am in the middle of making PR for allowing empty values in db. am just writing the testing instructions.

**Comment by richard67** [comment]
@brianteeman Could you check and test or review or both PR #27854 ?
